lock.go

```diff
@@ -0,0 +1,716 @@
+	// multiple locks with the given lockId, they will be unlocked in the
+	// reverse order in which they were created in (newest to oldest). For
+	// every lock that is unlocked, a LockStatus struct (representing the
+	// lock before it was unlocked) is returned. The order of these is the
+	// order in which they were unlocked.
+	//
+	// If an error is returned, it is safe and recommended for clients to
+	// retry until there is no error.
+	Unlock(lockId string) ([]LockStatus, error)
+
+	// Status returns the status of locks that match the provided filter.
+	// Fields with zero values in the Filter struct are ignored.
+	Status(f Filter) ([]LockStatus, error)
+
+	// Renew updates the TTL of all locks with the associated lockId. The
+	// new TTL for the locks will be the value of the argument provided.
+	//
+	// A LockStatus struct is returned for each lock that is renewed. The
+	// TTL field in each struct represents the new TTL. A client can use
+	// this information to ensure that all of the locks they created with
+	// the lockId have been renewed. If not all locks have been renewed,
+	// or if an error is returned, clients should assume that none of the
+	// locks are safe to use, and they should unlock the lockId.
+	Renew(lockId string, ttl uint) ([]LockStatus, error)
+}
+
+// mongoLocker implements the Locker interface with MongoDB.
+type mongoLocker struct {
+	coll *mgo.Collection
+}
+
+// NewMongoLocker creates a Locker that is backed by MongoDB.
+func NewMongoLocker(collection *mgo.Collection) Locker {
+	return &mongoLocker{
+		coll: collection,
+	}
+}
+
+func (l *mongoLocker) XLock(resourceName, lockId string, ld LockDetails) error {
+	selector := bson.M{
+		"resource":           resourceName,
+		"exclusive.acquired": false,
+		"shared.count":       0,
+	}
+
+	r := &resource{
+		Name:      resourceName,
+		Exclusive: lockFromDetails(lockId, ld),
+		Shared: sharedLocks{
+			Count: 0,
+			Locks: []lock{},
+		},
+	}
+
+	change := mgo.Change{
+		Update: r,
+		Upsert: true,
+	}
+
+	// One of three things will happen when we run this change (upsert).
+	// 1) If the resource exists and has any locks on it (shared or
+	//    exclusive), we will get a duplicate key error.
+	// 2) If the resource exists but doesn't have any locks on it, we will
+	//    update it to obtain an exclusive lock.
+	// 3) If the resource doesn't exist yet, it will be inserted which will
+	//    give us an exclusive lock on it.
+	_, err := l.coll.Find(selector).Apply(change, map[string]interface{}{})
+	if err != nil {
+		if mgo.IsDup(err) {
+			return ErrAlreadyLocked
+		}
+		return err
+	}
+
+	// Acquired lock.
+	return nil
+}
+
+func (l *mongoLocker) SLock(resourceName, lockId string, ld LockDetails, maxConcurrent int) error {
+	selector := bson.M{
+		"resource":           resourceName,
+		"exclusive.acquired": false,
+		"shared.locks.lockId": bson.M{
+			"$ne": lockId,
+		},
+	}
+	if maxConcurrent >= 0 {
```

Comment:
What happens if maxConcurrent is less than 0?

Reply:
https://github.com/mongo-go/lock/blob/d3390ceae528989b3b27c8ad81818e8f91289876/lock.go#L126

Reply:
Oops. It means "unlimited" shared locks allowed in the documentation of the interface